snuba/state/stateful_count.py

```diff
@@ -0,0 +1,60 @@
+from threading import Thread
+from time import time
+from typing import Any, Sequence, Tuple
+
+from arroyo.processing.strategies.dead_letter_queue import CountInvalidMessagePolicy
+from arroyo.processing.strategies.dead_letter_queue.policies.abstract import (
+    InvalidMessage,
+)
+from arroyo.types import TPayload
+
+from snuba.redis import redis_client
+
+
+class StatefulCountInvalidMessagePolicy(CountInvalidMessagePolicy[TPayload]):
+    """
+    An extension of the CountInvalidMessagePolicy which is able to save and load
+    the state of counted hits in Redis
+    """
+
+    def __init__(self, redis_hash_name: str, limit: int, seconds: int = 60) -> None:
+        self.__name = redis_hash_name
+        self.__seconds = seconds
+        super().__init__(limit, seconds, self._load_state())
+
+    def handle_invalid_message(self, e: InvalidMessage) -> None:
+        """
+        Asynchronously updates Redis hash with a new hit while
+        the base Policy handles the invalid message
+        """
+        Thread(target=self._add_to_redis).start()
+        super().handle_invalid_message(e)
```

Comment:
Where does your code actually check the current count to decide if this should throw or not?